### PR TITLE
Add U volume flag to chown source volumes

### LINKED ISSUE #7778: Option to use UIDs only once
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

Currently users of the same UID are always mapped to the same user on the host. User foo (UID 5) of imageA would be the same user as user bar (UID 5) of imageB. Since UIDs are given out sequentially, its fairly common that two arbitrary images would end up with users with the same ID.

When using bind mounts it might be desirable to chown the folders to unique UIDs (before starting the container), so that namespace separation can be maintained on the filesystem level too.

Podman could help to keep track of UIDs used in running containers and (optionally) make sure that every ID is only used once . Analog to the way how PIDs on the host work basically.

### PR DESCRIPTION
**Pull request by EduardoVega**
This will ensure that the uid,gid will be the same between the containers and volumes. The source volume will be chowned if the flag is specified to match the uid, gid of the container.

Fixes #7778 

Signed-off-by: Eduardo Vega <edvegavalerio@gmail.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
There is some fast chown code available in `containers/storage/drivers/chown` that can do a recursive chown with multiple threads, I believe.  We might want to take advantage of this.

**Comment by EduardoVega** [comment]
@rhatdan I am currently getting a CI issue because I have modified a vendor file (vendor/github.com/containers/buildah/pkg/parse/parse.go). It says `please sync the vendor.conf and commit all changes`, what's the correct way of doing this change? Thanks

**Comment by mheon** [comment]
Submit a PR of the change to Buildah, and then vendor the updated Buildah into Podman in this PR.

**Comment by EduardoVega** [comment]
Waiting for a new release of Buildah to met the dependencies for this PR.

**Comment by rhatdan** [comment]
You do not need to wait for a release.  You can just update the vendor of buildah here, based on major.  We will update the version of buildah before releasing a new version of podman.

**Comment by rhatdan** [comment]
@EduardoVega is this still a WIP pr?

**Comment by EduardoVega** [comment]
@rhatdan I think it is not since all buildah prerequisites were already present in podman.

**Comment by EduardoVega** [comment]
is there anything else I need to do ? or just wait until the PR is reviewed

**Comment by rhatdan** [comment]
Buildah should be released tomorrow, so we can move this ahead.

File: docs/source/markdown/podman-run.1.md
Comment (rhatdan) [inline docs/source/markdown/podman-run.1.md:1133]: This comment should mention that it will recursively change the UID of the files, and that this is potentially dangerous, and should be done with caution.

In SELinux we block certain volumes altogether from doing this.  Including $HOME.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:2144]: Not blocking for this PR, but we really should break this code out into a shared source with Buildah.  Should optimize for multi-threads like has been done for SELinux as well.  This procedure could take a long time.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:2149]: Message should mention chowning.

Comment (rhatdan) [reply libpod/container_internal_linux.go:2149]: This probably should be an error as well, since it would not be expected for this source to not exist.

Comment (rhatdan) [reply libpod/container_internal_linux.go:2149]: Finally the err will include the path, so dont put in in the message, otherwise the code will stutter.

**Review by rhatdan — APPROVED** [review]

**Comment by rhatdan** [comment]
@EduardoVega We would really to get this in by Friday's feature freeze.

File: docs/source/markdown/podman-run.1.md
Comment (EduardoVega) [reply docs/source/markdown/podman-run.1.md]: I have updated the docs and I have re used the relabel code for checking if the host path (source volume) can be modified.

**Comment by rhatdan** [comment]
This is not going to make the 3.0 release, Will have to wait until 3.1

**Comment by rhatdan** [comment]
@EduardoVega Is this still a WIP or do you want it reviewed?

**Comment by EduardoVega** [comment]
@rhatdan This is ready to be reviewed.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:2253]: This is fine for now, but this should be moved into Buildah so that it can take advantage of the list.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:2281]: Can you use
storage/pkg/homedir.Get() here?

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:2291]: I think you should use something like Realpath(path) on this. To make sure we are not being fooled.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:2253]: Nit, change the function to dangerousHostPath() and return true.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:2319]: Isn't the exact same code in Buildah?  Can you break it out so that Podman and Buildah can share the same Chowning code.

File: pkg/util/mountOpts.go
Comment (rhatdan) [inline pkg/util/mountOpts.go:117]: Why Not?  Does this happen automatically?  IE If I am running as a user, do the tmpfs get mounted as owned by that user?

File: test/e2e/run_volume_test.go
Comment (rhatdan) [inline test/e2e/run_volume_test.go:570]: Move to the top

**Review by rhatdan — APPROVED** [review]

File: libpod/container_internal_linux.go
Comment (EduardoVega) [reply libpod/container_internal_linux.go:2319]: Yes, this is the same code but sadly I did not create the method with capital letter so it can not be exported. I do not have any problem in sending another PR to Buildah to change that, so let me know the best way to proceed with this. thanks

Comment (rhatdan) [reply libpod/container_internal_linux.go:2319]: Let's do that.

File: libpod/container_internal_linux.go
Comment (EduardoVega) [reply libpod/container_internal_linux.go]: This has been moved to containers/common

File: libpod/container_internal_linux.go
Comment (EduardoVega) [reply libpod/container_internal_linux.go]: This has been changed in containers/common

File: libpod/container_internal_linux.go
Comment (EduardoVega) [reply libpod/container_internal_linux.go]: This has been changed in containers/common

File: libpod/container_internal_linux.go
Comment (EduardoVega) [reply libpod/container_internal_linux.go]: This has been changed in containers/common

File: pkg/util/mountOpts.go
Comment (EduardoVega) [reply pkg/util/mountOpts.go]: This was not happening automatically. To add correct ownership the uid and gid will be added to the mount options when the U flag is specified

File: test/e2e/run_volume_test.go
Comment (EduardoVega) [reply test/e2e/run_volume_test.go]: Done

**Comment by rhatdan** [comment]
Thanks @EduardoVega This is a nice new feature.
LGTM
@containers/podman-maintainers PTAL

**Comment by rhatdan** [comment]
@mheon @vrothberg PTAL

**Comment by rhatdan** [comment]
@EduardoVega We would love to have you present and demonstrate this at a monthly Podman users group, or write a blog on this feature.

File: libpod/container.go
Comment (mheon) [inline libpod/container.go:240]: Can you make this a more generic `Options []string` instead? It seems like we will have more of these later on, and the other volume types don't have per-option bools.

File: pkg/util/utils.go
Comment (mheon) [inline pkg/util/utils.go:698]: Suggest renaming to `IDtoolsToRuntimeSpec` (Capitalize the first D) - I think our linter checks for this one now. Might be it's not linting this directory?

**Comment by mheon** [comment]
Two suggestions, but on the whole this LGTM

**Comment by EduardoVega** [comment]
@rhatdan thanks and I think I could present it and also create a blog about this. I might need more details but I would love to do it.

File: pkg/util/utils.go
Comment (EduardoVega) [reply pkg/util/utils.go]: Thanks and done

File: libpod/container.go
Comment (EduardoVega) [reply libpod/container.go]: done

**Comment by mheon** [comment]
/lgtm
Nice work @EduardoVega

**Comment by TomSweeneyRedHat** [comment]
@EduardoVega very nice work, ty!  I too would love to have you present at a Podman community meeting.  They are on the first Tuesday of the month.  Next week is filled up, but the April one I'd love to have you present at.  However, we're holding that one later in the day than usual, 8:00 p.m. Eastern so the folks in China can attend.  So if you wanted to wait until the one in May which would be at 11:00 a.m. Eastern, that would work too.

**Comment by EduardoVega** [comment]
@TomSweeneyRedHat April one should work. I'm on Central America Time which is 1 hour behind of Eastern

**Comment by TomSweeneyRedHat** [comment]
@EduardoVega Awesome!  Would you mind sending me an e-mail (tsweeney@redhat.com) so I can touch base with you in mid-March with meeting info?

**Comment by EduardoVega** [comment]
@TomSweeneyRedHat sure.

**Pull request merged**
